### DISCUSSION TITLE
"Private" bot interactions . . .?

### DISCUSSION THREAD

**Discussion by humblecoder** [telegraf/telegraf#1735] [Q&A]
**NOTE:**
I realize this is a TELEGRAM issue rather than a TELEGRAFJS issue, but still curious about different keyboard/interaction implementations.

**GOAL:**
I'd like to facilitate private interactions between **ADMINS** and the bot.  I prefer the `actions`-based nature of `inlineKeyboard`, but I prefer the privacy of `keyboard`.  I've also noticed some bots simply taking users to PM.

**QUESTIONS:**
- For testing purposes, are there any good coding structures that allow a quick toggle between `keyboard` and `inlineKeyboard` that also preserves functionality?
- Can someone point me to any quality examples or best practices for admin bot interactions via PM?

**Comment by MKRhere** [comment]
What you want is not clear. Do you mean you want an interface for the admins of a group that's not visible to everyone? That's not possible. You can of course always link to bot private chat with a link or button. Could you clarify what you intend to achieve better?

**Comment by humblecoder** [comment]
**TL;DR**  Give admins ability to adjust custom room bot settings without showing to rest of chat, or being interrupted by scrolling chat.

-- 
Yes, I want private bot interactions for admins.  Something that no one else sees, and something that isn't affected by the main chat scrolling and potentially missing the prompt.

The bot has settings on a per-room basis, so I need to maintain context from the room from which they came.  As I mentioned, I've seen bots do PMing.  However, when I've attempted, say `switchToChat()` I receive errors that I need to share a person or room with whom to share information.  

I haven't fully researched the issue yet as I've been fixing other bugs.  But I thought I'd ask what **specific** solutions others have used to manage this.

**Reply by MKRhere** [reply]
Use a start parameter in the button URL, like `https://t.me/your_bot?start=group_${chat_id}`. And on the other side, you can handle it like this:

```TS
bot.start(ctx => {
  // ctx.startPayload // group_82173981
});
```

There's no other way to show an interface for only some members of the group, admins or not.

**Reply by humblecoder** [reply]
Alright, so what is the "high level overview" of what's occurring?  Are you saying I must spawn a separate group for every "private" conversation?  Or even worse, I must use separate bots to perform this "dance"?

I'm a bit confused.  I see bots do what I'm asking all the time.  Unfortunately they're not very forthcoming with details.

1. Admin (or whomever they allow) in group calls:  /changesettings
2. Bot presents button:  "Click here to chat . . . "
3. Click button and taken to private chat
4. Adjust desired settings alone in the room
5. Return to full group chat


What am I misunderstanding?

**Reply by MKRhere** [reply]
> Are you saying I must spawn a separate group?

> I must use separate bots?

No, just send the user to the same bot's own private chat.

**Reply by humblecoder** [reply]
Apologies for being "obtuse", but entering the private chat I still receive the "Share bot content . . .  Select a user or chat to share content via Bot".  How do I actually interact in the private room using the existing stages and workflow?

To be clear, I already have the bot working in the group chat.  I want to leave all current functionality "intact".  I simply want to take the conversation to the private chat **with appropriate context** so as not to "clutter" the main room.  What is the "least worst" way to accomplish this?

FYI, i'm currently using `launch()` via Telegraf to start the bot.  Not sure if that's relevant.  Also, this bot will be used in multiple rooms, so all `chat_id` settings and the like must be dynamic.

Again, I realize this is a TELEGRAM issue, but if there's knowledge here in the TELEGRAF community, I appreciate the input.
